Intro video script (#161)
buildpacks/rfcs: text/0000-intro-video-script.md

# Meta
[meta]: #meta
- Name: Intro video script
- Start Date: 2021-05-18
- Author(s): yaelharel
- RFC Pull Request: (leave blank)
- CNB Pull Request: (leave blank)
- CNB Issue: (leave blank)
- Supersedes: N/A

# Summary
[summary]: #summary

Improving documentation is a project priority for 2021.
This RFC is an attempt to create an easy and clear starting point for new users and contributors.

# Definitions
[definitions]: #definitions

N/A

# Motivation
[motivation]: #motivation

It is currently relatively challenging for new users and contributors to understand what buildpacks are and what is the purpose of our project.
It can be very useful to have a ~1 minute introductory video to provide high-level answers to these questions.
We can post this video on our website and share it on social media.
Hopefully this will help familiarize more people with our project.

# What it is
[what-it-is]: #what-it-is

### The script

> When was the last time you struggled to wrangle yet another Dockerfile? As an application developer, you should only have to focus on what you do best - writing code, without having to worry about your container build strategy. Buildpacks can help you with just that! They are a better approach to building container images for applications.

Author: I think opening with a question is more engaging and triggers the viewers' curiosity.

> When was the last time you struggled to wrangle yet another Dockerfile? As an application developer, you should only have to focus on what you do best - writing code, without having to worry about your container build strategy. Buildpacks can help you with just that! They are a better approach to building container images for applications.

Reviewer: generally I agree, but I worry this question might alienate people. What if they haven't struggled with a Dockerfile? What if they like Dockerfile and don't understand why buildpacks are better? I guess I also worry that putting so much emphasis on Dockerfile upfront has risks. I also might be overthinking it.

> When was the last time you struggled to wrangle yet another Dockerfile? As an application developer, you should only have to focus on what you do best - writing code, without having to worry about your container build strategy. Buildpacks can help you with just that! They are a better approach to building container images for applications.

Author: Interesting point of view.
@jromero, @natalieparellano, what do you think?

> When was the last time you struggled to wrangle yet another Dockerfile? As an application developer, you should only have to focus on what you do best - writing code, without having to worry about your container build strategy. Buildpacks can help you with just that! They are a better approach to building container images for applications.

Reviewer 2: I may also be overthinking this - but a very common people question ask is why buildpacks over dockerfiles (this came up in our user research and kubecon for eg.) More often than not, people don't struggle with Dockerfiles, mostly because it is really easy to write one (albient it is hard to write a **good** Dockerfile). 

Although not a good representation of our user base, if you look at the comments (we should also take this with a grain of salt since it is hackernews) here -> https://news.ycombinator.com/item?id=26041099, most newcomers to buildpacks fail to see why they would want to transition from Dockerfiles to buildpacks. 

I think people need a context like Joe's recent dockerfile article (https://jkutner.github.io/2021/04/26/write-good-dockerfile.html) to understand why dockerfiles can be a leaky abstraction and why they are bad. 

Maybe something like - 

> Have you ever wished your apps could be containerized automatically without having to copy paste random Dockerfile snippets in every project? As an application developer, you should only have to focus on what you do best - writing code; without having to worry about things like image security, creating minimal images or carefully crafting those Dockerfile instructions for proper caching. Buildpacks offer all of that and more!

> When was the last time you struggled to wrangle yet another Dockerfile? As an application developer, you should only have to focus on what you do best - writing code, without having to worry about your container build strategy. Buildpacks can help you with just that! They are a better approach to building container images for applications.

Author: @samj1912, thanks a lot for your comment!!!
I really like your idea!

The only two concerns that I have are:
- Is it too detailed? Probably we don't catch all of the pain points that programers have with Dockerfiles.
I don't want viewers to not try buildpacks because we didn't mentioned their Dockerfile problem (while "When was the last time you struggled to wrangle yet another Dockerfile?" kind of covers "all of the problems").
- I think that our script is a bit long for a 1-minute video and by editing these few sentences, we're even make it longer. Maybe we should delete parts from the rest of the script? If so, I'm not sure what.

> When was the last time you struggled to wrangle yet another Dockerfile? As an application developer, you should only have to focus on what you do best - writing code, without having to worry about your container build strategy. Buildpacks can help you with just that! They are a better approach to building container images for applications.

Reviewer 2: > * Is it too detailed? Probably we don't catch all of the pain points that programers have with Dockerfiles.
>   I don't want viewers to not try buildpacks because we didn't mentioned their Dockerfile problem (while "When was the last time you struggled to wrangle yet another Dockerfile?" kind of covers "all of the problems").

I agree that we wont be able to capture all issues with Dockerfile if we get too much into the specifics but on the other hand I fear that a statement like "When was the last time you struggled to wrangle yet another Dockerfile?" maybe too abstract/generic. Our users have complained about the docs homepage being too high-level. I was hoping we could highlight some common issues that may appeal to both the app developer and operator personas, but I don't have a strong opinion on this given the time limitations.

> When was the last time you struggled to wrangle yet another Dockerfile? As an application developer, you should only have to focus on what you do best - writing code, without having to worry about your container build strategy. Buildpacks can help you with just that! They are a better approach to building container images for applications.

Author: Yeah, I'm not sure what is the right approach.
This intro video is supposed to be very basic, but its purpose is also to make people want to try buildpacks, so maybe we should make it more specific. I'm not sure.
@sampeinado, would love to get your opinion.

> When was the last time you struggled to wrangle yet another Dockerfile? As an application developer, you should only have to focus on what you do best - writing code, without having to worry about your container build strategy. Buildpacks can help you with just that! They are a better approach to building container images for applications.

Reviewer 3: I'd say start with a simple statement rather a question that gets more to the core value proposition. This is kind of my intuition from sitting through a lot of startup pitches. Analogies/metaphors are good for helping people understand, but sometimes can get a little muddle. Better for discussion than a short pitch. 

For instance: 
Containerization technologies are making growing and scaling a software application from idea to a large deployment easier. Buildpacks make it easy to get from source code to container image.

> When was the last time you struggled to wrangle yet another Dockerfile? As an application developer, you should only have to focus on what you do best - writing code, without having to worry about your container build strategy. Buildpacks can help you with just that! They are a better approach to building container images for applications.

Author: I've made a few changes thanks to your feedback. Please see https://github.com/buildpacks/rfcs/pull/161/commits/fab2b12a7f5048e25a18d35d5240198f5f7e7b73.

> As an application developer, you should focus on what you do best - writing code, without having to worry about image security, optimizing container images, or your container build strategy. How much time have you spent struggling to wrangle yet another Dockerfile? Copying and pasting random Dockerfile snippets into every project? Buildpacks can help you with just that! They are a better approach to building container images for applications.

Reviewer 4: Do we care about addressing another target audience or this is just a video for app developers?

Author: The purpose of this video is to make a clear and short introduction to what buildpacks are.
I think that the easiest way to address it is from the app developers perspective.

Reviewer 5: My dream: one day we have a separate video for platform operators :)

Author: @natalieparellano, it will be great!

> As an application developer, you should focus on what you do best - writing code, without having to worry about image security, optimizing container images, or your container build strategy. How much time have you spent struggling to wrangle yet another Dockerfile? Copying and pasting random Dockerfile snippets into every project? Buildpacks can help you with just that! They are a better approach to building container images for applications.

Reviewer 5: ```suggestion
As an application developer, you should focus on what you do best - writing code, without having to worry about image security, optimizing container images, or your container build strategy. How much time have you spent struggling to wrangle yet another Dockerfile? Copying and pasting random Dockerfile snippets into every project? Buildpacks can help you do better! They are a better approach to building container images for applications.
```

Reviewer 5: (Now we have "better" two sentences in a row... it's just that before I thought buildpacks would help me copy/paste my Dockerfiles)

Reviewer 5: Edited my suggestion :)

> As an application developer, you should only focus on what you do best - writing code, without having to worry about image security, creating minimal images, and your container build strategy in general. But how much time have you wasted struggled to wrangle yet another Dockerfile? Copy-pasting random Dockerfile snippets in every project? Buildpacks can help you with just that! They are a better approach to building container images for applications.

Reviewer 6: ```suggestion
As an application developer, you should focus on what you do best - writing code, without having to worry about image security, optimizing container images, or your container build strategy. How much time have you spent struggling to wrangle yet another Dockerfile? Copying and pasting random Dockerfile snippets into every project? Buildpacks can help! They are a better approach to building container images for applications.
```

Reworded some of this to make it clearer.

Author: Please see https://github.com/buildpacks/rfcs/pull/161/commits/9ee41ece3bdfce2666be374d3ae42b82a6becd91.

> Buildpacks are programs that transform source code into container images by analyzing the code and determining the best way to build it. They can provide dependencies, configuration, run compilation, and much more! A buildpack can be written by anyone, and shared with everyone! Google, Heroku/Salesforce, and the Paketo project provide buildpacks for a wide variety of language families, and you can discover even more by searching the buildpack registry. With buildpacks, you can use best-in-class container build strategies from industry experts with ease.

Reviewer: "Heroku/Salesforce" won't be easy to say. We can make this "Heroku, Salesforce, and ..."

Author: Please see https://github.com/buildpacks/rfcs/pull/161/commits/084534fe1e7cf27c4589ea6215633518a0a5b9d6.

> Buildpacks are programs that transform source code into container images by analyzing the code and determining the best way to build it. They can provide dependencies, configuration, run compilation, and much more! A buildpack can be written by anyone, and shared with everyone! Google, Heroku, Salesforce, and the Paketo project provide buildpacks for a wide variety of language families, and you can discover even more by searching the buildpack registry. With buildpacks, you can use best-in-class container build strategies from industry experts with ease.

Reviewer 2: The part mentioning vendors here and the one below should be combined?

Author: Yeah, probably it's better to mention the list of vendors only once in the video.
Do you think we should keep editing it?
I think that it's ok to discuss it with the person that will work on the video, but I'm also ok with making a few more changes in order to make it clearer.

> Buildpacks are programs that transform source code into container images by analyzing the code and determining the best way to build it. They can provide dependencies, configuration, run compilation, and much more! A buildpack can be written by anyone, and shared with everyone! Google, Heroku, Salesforce, and the Paketo project provide buildpacks for a wide variety of language families, and you can discover even more by searching the buildpack registry. With buildpacks, you can use best-in-class container build strategies from industry experts with ease.

Reviewer 6: ```suggestion
Buildpacks are programs that transform source code into container images by analyzing the code and determining the best way to build it. They can provide dependencies, configuration, run compilation, and much more! A buildpack can be written by anyone, and shared with everyone! Google, Heroku, and the Paketo project provide buildpacks for a wide variety of language families, and you can discover even more by searching the buildpack registry. With buildpacks, you can use best-in-class container build strategies from industry experts with ease.
```

AFAIK, Salesforce doesn't provide any publically available buildpacks. It's all Heroku 😜

Author: Oops, you're right. Thanks!

Author: Please see https://github.com/buildpacks/rfcs/pull/161/commits/9ee41ece3bdfce2666be374d3ae42b82a6becd91.

> Even more so, the Cloud Native Buildpacks project provides a specification for buildpack authors and platforms that ensures builds are fast, secure, customizable, and can run on any cloud. It is a Cloud Native Computing Foundation project with many existing integrations and many more to come.

Reviewer 3: I think it'd be good to add back in some of the copy for social proof in this paragraph:

> Once applications are in containers there is an incredible community of tools and products that can help get your app online and support its growth. Heroku, Hashicorp Waypoint, VMWare Tanzu, Spaceship, CircleCI, and Google Cloud Run all support using buildpacks to make container images.

Author: Thank you for your comment!
As I wrote https://github.com/buildpacks/rfcs/pull/161/files#r637173592, my top concern is that it'll make our script too long.
We're trying to create a very basic 1 minute video and I'm afraid that even before the additions that you and Sam suggested, our script is too long.

Reviewer 6: My 2c: One thing to keep in mind is that this script will be subjective to additional changes throughout the next steps of productions. I wouldn't be too hung up on specific word counts. If the additions add value they may be worth a few more seconds of air time. It's probably better to have too much to cut out in the next step as oppose to not enough material.

Author: @thephw, I took @jromero advice and added your sentence as is for now.

> Even more so, the Cloud Native Buildpacks project provides a specification for buildpack authors and platforms that ensures builds are fast, secure, customizable, and can run on any cloud. It is a Cloud Native Computing Foundation project with many existing integrations. Platforms such as Google Cloud Run, Hashicorp Waypoint, Heroku, Spaceship, and VMWare Tanzu, just to name a few, all support using buildpacks to make container images.

Reviewer 5: ```suggestion
Even more so, the Cloud Native Buildpacks project provides a specification for buildpack authors and platforms that ensures builds are fast, secure, customizable, and can run on any cloud. It is a Cloud Native Computing Foundation project with many existing integrations. Platforms that use buildpacks to make container images include Google Cloud Run, Hashicorp Waypoint, Heroku, Spaceship, and VMWare Tanzu, just to name a few.
```

Author: Please see https://github.com/buildpacks/rfcs/pull/161/commits/cdb19a16889aafc1a54dd6582212684c60732b74.

> Even more so, the Cloud Native Buildpacks project provides a specification for buildpack authors and platforms that ensures builds are fast, secure, customizable, and can run on any cloud. It is a Cloud Native Computing Foundation project with many existing integrations. Platforms such as Google Cloud Run, Hashicorp Waypoint, Heroku, Spaceship, and VMWare Tanzu, just to name a few, all support using buildpacks to make container images.

Reviewer 5: ```suggestion
Even more, the Cloud Native Buildpacks project provides a specification for buildpack authors and platforms that ensures builds are fast, secure, customizable, and can run on any cloud. It is a Cloud Native Computing Foundation project with many existing integrations. Platforms such as Google Cloud Run, Hashicorp Waypoint, Heroku, Spaceship, and VMWare Tanzu, just to name a few, all support using buildpacks to make container images.
```

Author: Please see https://github.com/buildpacks/rfcs/pull/161/commits/cdb19a16889aafc1a54dd6582212684c60732b74.

You are invited to join our community. Try buildpacks with your favorite language today.

# How it Works
[how-it-works]: #how-it-works

After this RFC will get accepted and merged, we will start working on the actual video. 

# Drawbacks
[drawbacks]: #drawbacks

> N/A

Reviewer: I think there's a drawback that maintaining the video is harder than editing text.

Author: I'll add it, although I think that we won't need to change it anytime soon because it's very basic.

Author: Please see https://github.com/buildpacks/rfcs/pull/161/commits/084534fe1e7cf27c4589ea6215633518a0a5b9d6.

> Even more so, the Cloud Native Buildpacks project provides a specification for buildpack authors and platforms that ensures builds are fast, secure, customizable, and can run on any cloud. It is a Cloud Native Computing Foundation project with many existing integrations and many more to come. Once applications are in containers there is an incredible community of tools and products that can help get your app online and support its growth. Heroku, Hashicorp Waypoint, VMWare Tanzu, Spaceship, CircleCI, and Google Cloud Run all support using buildpacks to make container images.

Reviewer 6: I'm not sure I'm keen on name dropping just a few integrations or platforms. For buildpack providers it doesn't seem like a huge deal since we already "suggest" them via things like `pack builders suggest`. For platforms and integrations, I'm curious as to why only this subset was selected.

Concerns:

* It seems subjectively bias to only call out a subset of integrations/platforms.
* It increases the staleness factor as platforms may come and go.

That said, I tried to identify the integrations as platforms and alphabetize them.

```suggestion
Even more so, the Cloud Native Buildpacks project provides a specification for buildpack authors and platforms that ensures builds are fast, secure, customizable, and can run on any cloud. It is a Cloud Native Computing Foundation project with many existing integrations. Platforms such as Google Cloud Run, Hashicorp Waypoint, Heroku, Spaceship, and VMWare Tanzu, just to name a few, all support using buildpacks to make container images.
```

Author: I'm going to make this change but I have to say that I'm not sure the last sentence (starting with "Platforms such as...") is necessary.
Any thoughts?

# Alternatives
[alternatives]: #alternatives

N/A

# Prior Art
[prior-art]: #prior-art

N/A

# Unresolved Questions
[unresolved-questions]: #unresolved-questions

> * How many words can a typical 1-minute video script include? Is the current script too long/short?

Reviewer 4: It will be hard to say all of this text within 1 minute. I think that's ok.

Author: Yeah, you're right. According to @thephw contractor, there can be around 125 words per minute and now we're around 200 words.
But since this is a draft script and we'll continue to edit it while working on the actual video, I think it's ok.
Anyway, we're aiming that the video won't take much more than 1 minute.
